common/css.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <compiler.h>
+#include <css.h>
+#include <debug.h>
+#include <error.h>
+#include <stdint.h>
+#include <util.h>
+
+/*
+ * These generic functions provide sane default implementations of the CSS API,
+ * so platforms may incrementally implement power domain control functions.
+ */
+uint8_t __weak
+css_get_css_state(void)
+{
+	/* Assume the CSS is always on. */
+	return POWER_STATE_ON;
+}
+
+uint8_t __weak
+css_get_cluster_count(void)
+{
+	/* Assume the CSS contains a single cluster with a single core. */
+	return 1;
+}
+
+uint8_t __weak
+css_get_cluster_state(uint8_t cluster __unused)
+{
+	/* Assume present clusters/cores are always on. */
+	if (cluster >= css_get_cluster_count())
+		return POWER_STATE_OFF;
+
+	return POWER_STATE_ON;
+}
+
+uint8_t __weak
+css_get_core_count(uint8_t cluster __unused)
+{
+	/* Assume the CSS contains a single cluster with a single core. */
+	return 1;
+}
+
+uint8_t __weak
+css_get_core_state(uint8_t cluster, uint8_t core)
+{
+	/* Assume present clusters/cores are always on. */
+	if (cluster >= css_get_cluster_count())
+		return POWER_STATE_OFF;
+	if (core >= css_get_core_count(cluster))
+		return POWER_STATE_OFF;
+
+	return POWER_STATE_ON;
+}
+
+/*
+ * There should usually be no reason to override this weak definition. It
+ * correctly implements the algorithm specified in the CSS API using the
+ * lower-level core state API.
+ */
+uint8_t __weak
+css_get_online_cores(uint8_t cluster)
+{
+	uint8_t cores;
+	uint8_t mask = 0;
+
+	assert(cluster < css_get_cluster_count());
+
+	cores = css_get_core_count(cluster);
+	for (uint8_t core = 0; core < cores; ++core) {
+		if (css_get_core_state(cluster, core) != POWER_STATE_OFF)
+			mask |= BIT(core);
+	}
+
+	return mask;
+}
+
+int __weak
+css_set_css_state(uint8_t state)
+{
+	/* Reject any attempts to change CSS, cluster, or core power states. */
```

Comment:
This isn't TODO (i.e. I don't plan to change this file in the future). This is the safe thing to do when there is no platform hardware support.

Reply:
Then this needs further explanation in the comments. This looks like a partial implementation.

Reply:
It _is_ a partial implementation. It's the "what does the API call do if we don't know how the hardware works?" implementation. That's in a comment at the top of the file:

```
/*
 * These generic functions provide sane default implementations of the CSS API,
 * so platforms may incrementally implement power domain control functions.
 */
```

Do you want me to expand that comment?

Reply:
Default does not mean partial. And they _should_ be default implementations rather than partial implementations. What I recommend doing (and feel free to add your input, @agom94) is changing it to read something like: 
"Changing CSS, cluster, or core power states are forbidden by default without platform specific implementation."